cmd/compose/cp.go

```diff
@@ -64,8 +64,10 @@ func copyCommand(p *projectOptions, backend api.Service) *cobra.Command {
 	}
 
 	flags := copyCmd.Flags()
-	flags.IntVar(&opts.index, "index", 1, "Index of the container if there are multiple instances of a service [default: 1].")
+	flags.IntVar(&opts.index, "index", 0, "Index of the container if there are multiple instances of a service [default: 1 when copying from a container].")
```

Comment:
Not a huge fan of specifying a default value that we change depending of the context of the `cp` command. if you have better idea to do it, let me know

Reply:
Perhaps remove the `[default: 1 when copying from a container].` altogether? (note that Cobra already prints defaults, so in most cases, there's no need to manually put it in the description)

```suggestion
	flags.IntVar(&opts.index, "index", 0, "Index of the container if there are multiple instances of a service .")
```

As to the flag itself; are indexes guaranteed to start with `1`? (i.e., could a container be removed, and it starts with `2` ?)

The "when copying from a container" is a bit confusing, as compose works with "services", correct? (in other words; I don't think there's an option to specify a _container_ - for that you'd use `docker cp` or `docker container cp`.

Overall I'm wondering a bit if we should have this option; my train of thought here is that;

- by default, we should assume all instances of a service to be identical
- compose to interact with a compose "service" (so consider all to be identical)
- it there _IS_ a need to target a specific instance (which _should be_ an exception to the rule); perhaps users should use `docker cp` / `docker container cp` for the "I know what I'm doing" situation.

Reply:
we need to double check `index` is aligned wit `compose exec --index`, and that we didn't duplicated the logic to select a target container here so that we enforce an homogeneous behavior

when copying from a container indeed only make sense with a single container as target. Either we should REQUIRE `index` to be set when multiple container match, or set "first indexed in the available ones" as default (which might not be `1`) - but, again, should behave the same as `compose exec`

> it there IS a need to target a specific instance (which should be an exception to the rule); perhaps users should use docker cp / docker container cp for the "I know what I'm doing" situation.

While it obvious they can just run `docker exec <container>` as name is predicable with compose, for some reason (?) `docker-compose exec` was introduced and adopted by many, probably because users prefer to rely on the same tool for everything (when you get a hammer...). So, even with _containers_ as target, a compose command makes sense to them. Or maybe this helps writing scripts that do not depend on te project name (included in container name) set by project dir?

Reply:
(let me "unresolve" so that it's not hidden)

Reply:
`docker compose exec  <containername>` doesn't make sense to me, the whole verbage used by compose is service-centric. Use of `--index` is unpleasant but I can't imagine a workaround. I'd just expect it's not required for service running 1 replica and user don't have to bother with this BUT when they try to diagnose a distributed-system issue with multiple replicas. But then, obviously, using plain container name with `docker exec` would be a natural option, and maybe we should deprecate `--index` support and suggest user to better use `docker exec` for this purpose. Or maybe this is fine we keep this as legacy/usability, as we - at least - share runExec implementation code!